src/Lean/Data/Name.lean

```diff
@@ -65,26 +65,50 @@ def isSuffixOf : Name → Name → Bool
   | num p₁ n₁ _, num p₂ n₂ _ => n₁ == n₂ && isSuffixOf p₁ p₂
   | _,           _           => false
 
-def lt : Name → Name → Bool
-  | anonymous,   anonymous   => false
-  | anonymous,   _           => true
-  | num p₁ i₁ _, num p₂ i₂ _ => lt p₁ p₂ || (p₁ == p₂ && i₁ < i₂)
-  | num _ _ _,   str _ _ _   => true
-  | str p₁ n₁ _, str p₂ n₂ _ => lt p₁ p₂ || (p₁ == p₂ && n₁ < n₂)
-  | _,           _           => false
-
-def quickLtAux : Name → Name → Bool
-  | anonymous, anonymous   => false
-  | anonymous, _           => true
-  | num n v _, num n' v' _ => v < v' || (v = v' && n.quickLtAux n')
-  | num _ _ _, str _ _ _   => true
-  | str n s _, str n' s' _ => s < s' || (s = s' && n.quickLtAux n')
-  | _,         _           => false
+def cmp : Name → Name → Ordering
+  | anonymous,   anonymous   => Ordering.eq
+  | anonymous,   _           => Ordering.lt
+  | _,           anonymous   => Ordering.gt
+  | num p₁ i₁ _, num p₂ i₂ _ => 
+    match cmp p₁ p₂ with  
+    | Ordering.eq => compare i₁ i₂
+    | ord => ord
```

Comment:
Should we have a function for this? In Haskell, this is simply the `Monoid` instance for `Ordering`, but we'd have to be a bit more careful with strictness...

Reply:
so I guess we could have a `combine : Ordering -> Thunk Ordering -> Ordering`, but is it really so much easier?

Reply:
sometimes, I feel like the 3-line solution is ok.

Reply:
Yeah, it's not too bad :)

Reply:
I would imagine this pattern is common enough I would probably find myself reaching for something like a default `Ord` instance for products to handle it (assuming the comparisons are inlined at least... perhaps a variadic macro could work too?)
```lean
instance [Ord α] [Ord β] : Ord (α × β) where
  compare | (a₁, b₁), (a₂, b₂) =>
            match compare a₁ a₂ with
            | Ordering.eq => compare b₁ b₂
            | ord => ord

#eval compare (1,2,3) (1,2,4)
```

Reply:
Ah but the first use in that example uses `cmp` and the second usage is to `compare`... so perhaps this is less applicable here.

Reply:
Yes, outside of recursive `Ord` instances, just using pairs seems like a nice solution

Reply:
The `import` time improved a bit after I added `@[inline]` at
```lean
@[inline] def compareOfLessAndEq {α} ...
```
but it is still slower.